README.rst
.. image:: https://img.shields.io/pypi/v/giraffez.svg
     :target: https://pypi.python.org/pypi/giraffez
     :alt: PyPi
.. image:: https://img.shields.io/badge/license-Apache%202-blue.svg
     :target: https://www.apache.org/licenses/LICENSE-2.0
     :alt: License

|

.. image:: https://raw.githubusercontent.com/capitalone/giraffez/master/docs/logo.png

Giraffez is a client library for `Teradata <http://www.teradata.com/>`_, made to be fast and user-friendly. It shares many features with the official Teradata tools, but makes them available in an easy-to-use command-line interface and Python API.


Features
########

- `Command-line tool <https://capitalone.github.io/giraffez/command-line.html>`_
- `Python API <https://capitalone.github.io/giraffez/api.html#giraffez-modules>`_ - expressive, powerful, and great for `working with other packages <https://capitalone.github.io/giraffez/api.html#working-with-other-packages>`_ (e.g. `pandas <http://pandas.pydata.org>`_)
- `Execute <https://capitalone.github.io/giraffez/command-line.html#cmd>`_ SQL statements using the Teradata CLIv2 library -- performing ad hoc queries and commands
- `Insert <https://capitalone.github.io/giraffez/command-line.html#insert>`_ into existing Teradata tables using the CLIv2 drivers
- `Archive <https://capitalone.github.io/giraffez/command-line.html#archiving>`_ Teradata tables to save space. These can easily be `transformed to flatfiles <https://capitalone.github.io/giraffez/command-line.html#fmt>`_ or `loaded <https://capitalone.github.io/giraffez/command-line.html#load>`_ back into Teradata
- `BulkExport <https://capitalone.github.io/giraffez/command-line.html#export>`_ from Teradata to a flatfile or archival format, retaining data type information
- `BulkLoad <https://capitalone.github.io/giraffez/command-line.html#load>`_ into existing Teradata tables using the Teradata PT Update (MLOAD)
- `Config <https://capitalone.github.io/giraffez/command-line.html#config>`_ - manage connection information and credentials securely using the ``giraffez config`` and ``giraffez secret`` commands
- `Interactive shell <https://capitalone.github.io/giraffez/command-line.html#shell>`_ for executing Teradata queries -- much like BTEQ, but with a friendly interface
- `Run <https://capitalone.github.io/giraffez/command-line.html#run>`_ job files in YAML format

Links
#####

- `Documentation <https://capitalone.github.io/giraffez>`_

Quick Install
#############

::

  $ pip install giraffez

See the `setup instructions <https://capitalone.github.io/giraffez/intro.html#giraffez-setup>`_ for requirements and details on how to configure your environment.

Contributors
############

We welcome Your interest in Capital One’s Open Source Projects (the
"Project"). Any Contributor to the Project must accept and sign an
Agreement indicating agreement to the license terms below. Except for
the license granted in this Agreement to Capital One and to recipients
of software distributed by Capital One, You reserve all right, title,
and interest in and to Your Contributions; this Agreement does not
impact Your rights to use Your own Contributions for any other purpose

`Sign the Agreement <https://docs.google.com/forms/d/e/1FAIpQLSfwtl1s6KmpLhCY6CjiY8nFZshDwf_wrmNYx1ahpsNFXXmHKw/viewform>`_

Code of Conduct
###############

This project adheres to the `Open Code of Conduct <https://developer.capitalone.com/single/code-of-conduct/>`_ By participating, you are
expected to honor this code.

----

Teradata is a registered trademark of Teradata Corporation



CHANGELOG
giraffez Changelog
======================

Version 2.1.0
-------------
Nov 9 2018

- Add feature for specifying logon mechanism (enables LDAP and other logon
functionality)
- Fix a few more Python 3.7 affected calls to PyUnicode_AsUTF8
- Resolves #63, unfortunately the bug is in the TTU and cannot be fix in
giraffez

Version 2.0.24
-------------
Sep 11 2018

- Fix #62

Version 2.0.23
-------------
Aug 13 2018

- Fix #50

Version 2.0.22
-------------
Aug 13 2018

- Fix #53

Version 2.0.21
-------------
Mar 7 2018

- Fix #49

Version 2.0.20
-------------
Feb 1 2018

- Merge #43

Version 2.0.19
-------------
Oct 13 2017

- Fix row string serialization attaching extra delimiters to rows when NULL

Version 2.0.18
-------------
Oct 10 2017

- Fix memory leak in conversion from decimal to pyfloats
- Fix ignored setting coerce_floats when set to false and using either list or dict as results output

Version 2.0.17
-------------
Sep 20 2017

- Fix MANIFEST.in to include all .c files (fixes #41)

Version 2.0.16
-------------
Sep 19 2017

- Fix giraffez fmt command
- Fix python 2.7 str conversion to int and float 

Version 2.0.15
-------------
Sep 18 2017

- No longer use extremely broken PyOS_stricmp

Version 2.0.14
-------------
Sep 18 2017

- Fix columns not set in BulkLoad in `from_file`
- Use PyOS_stricmp instead of strcmp for case-insensitive comparison of
  column_list when setting schema for BulkLoad
- Raise exception on duplicate columns in input for BulkLoad

Version 2.0.13
-------------
Sep 12 2017

- Decimal 32-bit fix on OSX

Version 2.0.12
-------------
Sep 12 2017

- Fix missing pointer position in C buffer_writef

Version 2.0.11
-------------
Sep 8 2017

- Fix python2.7 issue with column names being converted from unicode to str

Version 2.0.10
-------------
Sep 7 2017

- Fix incorrect method name on command-line insert

Version 2.0.9
-------------
Sep 6 2017

- Attempted fix of regression introduced in previous commit to cleanup
  C code creating columns.  Some values must not be initialized properly
  as this changed caused free() invalid pointer errors.

Version 2.0.8
-------------
Aug 31 2017

- Fix regression in behavior with dict return type where the column names
  are not preferring the alias where provided.

Version 2.0.7
-------------
Aug 29 2017

- Fix an issue with Context which prevented the use of arguments to
  the __init__ of a subclass

Version 2.0.6
-------------
Aug 25 2017

- Fix error caused by non-ascii utf-8 being cut short with new CHAR
  handling logic.
- Add context guards on classes that should use with context to ensure
  connections are properly closed.

Version 2.0.5
-------------
Aug 22 2017

- Remove unused methods/properties from BuilkExport and BulkLoad API for
  null/delimiter (currently only used on 'to_str()' and 'from_file')

Version 2.0.4
-------------
Aug 18 2017

- More MSVC compatibility fixes

Version 2.0.3
-------------
Aug 17 2017

- Fix missing macro in older versions of Python

Version 2.0.2
-------------
Aug 16 2017

- Fix missing import 'message_write_default'

Version 2.0.1
-------------
Aug 16 2017

- Fix missing include in MANIFEST.in

Version 2.0.0
-------------
Aug 10 2017

- Improvement: New C serializer (faster MLoad encoding)
- Improvement: TeradataCmd API, there is now only one execute method
  that handles all cases (single statement, multiple statement (sync),
  multiple statement (async), file input)
- Improvement: TeradataMLoad API, no longer need to call finish method
- Improvement: TeradataExport API, convenient methods for returning
  different result types (dict, list, and str)
- Fix #37

Version 1.1.9
-------------
Jul 6 2017

- Fix #34

Version 1.1.8
-------------
Feb 28 2017

- Fix #26

Version 1.1.7
-------------
Dec 16 2016

- Fix setup.py library paths for Darwin

Version 1.1.6
-------------
Dec 13 2016

- Fixes #23, related to a major bug in how giraffez.Cmd established a
  session causing significant performance problems and exceptions
- Disable parallel builds in setup.py causing intermittent build
  failures
- Fix missing import in mload.py

Version 1.1.4
-------------
Nov 29 2016

- Fix missing/incorrect keyword arguments being passed to giraffez.Config
- Initial macosx build support

Version 1.1.3
-------------
Nov 18 2016

- Fixes #18

Version 1.1.2
-------------
Nov 14 2016

- Fix bug where GIRAFFE_NOT_FOUND global variable is overwritten by imports

Version 1.1.1
-------------
Nov 14 2016

- Fix `giraffez fmt` command
- Add `--header` option to `giraffez fmt`

Version 1.1.0
-------------
Nov 14 2016

- Restructure/consolidate C files
- File text parsing now uses the Python csv module (#15)
- Decimal type now returns as giraffez.Decimal (only in giraffez.Cmd)
- Fix reader modes in Python 3 when using gzip on non-archive output
- Rename some internal types for consistency

Version 1.0.5
-------------
Oct 5 2016

- Fix Python3 incompatible raw_input
- Add new methods to Row/Result to replace to_json
- Fix Export overwriting the same column names when using dict output

Version 1.0.4
-------------
Sep 5 2016

- Improve build tools and add graceful shutdown (no longer registered by
  default)
- Fix MOD_DEF macros not working in MSVC 10 compiler
- Add missing import in giraffez/parser.py
- Fix decimal64 overflow on MSVC

Version 1.0.3
-------------
Sep 1 2016

- All C extensions now required for successful build

Version 1.0.2
-------------
Sep 1 2016

- Fixes #3

Version 1.0.1
-------------
Aug 31 2016

- Fixes #1

Version 1.0
-----------
Aug 11 2016

- Initial commit


giraffez/types.py
# -*- coding: utf-8 -*-
#
# Copyright 2016 Capital One Services, LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import datetime
import math
import struct

from .constants import *
from .errors import *

from .fmt import safe_name
from .logging import log

from ._compat import *


__all__ = ['Column', 'Columns', 'Date', 'Decimal', 'Time', 'Timestamp', 'Row']


DATE_FORMATS = [
    "%m/%d/%Y",
    "%Y/%m/%d",
    "%Y-%m-%d",
    "%Y-%m-%d %H:%M:%S",
    "%Y%m%d",
]

TIME_FORMATS = {
    8: "%H:%M:%S",
    15: "%H:%M:%S.%f",
    19: "%Y-%m-%d %H:%M:%S",
    22: "%Y-%m-%d %H:%M:%S.%f",
    26: "%Y-%m-%d %H:%M:%S.%f",
    32: "%Y-%m-%d %H:%M:%S.%f"
}


class Column(object):
    """
    An object containing the column information used by Teradata for
    encoding, decoding, and identifying data.

    Implements the :code:`__eq__` magic method for comparison with other instances
    (validating that each field holds the same data), and the :code:`__str__` and
    :code:`__repr__` magic methods for coersion to string values and more pleasant
    visual presentation.

    **Fields:**

    name
        The original name of the column in the table it belongs to
    alias
        The alias of the column in the query
    title
        The title of the column in the query
    type
        Numeric value used by Teradata to indicate column type
    length
        Length in bytes that the field's values occupy
    precision
        Precision of floating-point numeric columns
    scale
        Scale of floating-point numeric columns
    filler
        Indicator that the column belongs to the target table, but
        will not be used in the current operation
    """

    def __init__(self, column, filler=False):
        if isinstance(column, tuple):
            if len(column) in {5, 7}:
                column = {"name": column[0], "type": column[1], "length": column[2], "precision": column[3], "scale": column[4]}
            elif len(column) == 8:
                column = {"name": column[0], "type": column[1], "length": column[2], "precision": column[3], "scale": column[4], "nullable": column[5], "default": column[6], "format": column[7]}
            else:
                raise GiraffeTypeError("Column information is not valid")
        self.original_name = ensure_str(column.get("name")).strip('"').lower()
        self.alias = (column.get("alias", "")).strip('"').lower()
        self.title = (column.get("title", "")).strip('"').lower()
        self.type = column.get("type")
        if self.type in cmd_type_map:
            self.tpt_type = self.type
            self.type = cmd_type_map.get(self.type)
        else:
            self.tpt_type = tpt_type_map.get(self.type)
        self.gd_type = gd_type_map.get(self.type, None)
        if self.type is None or self.gd_type is None:
            raise GiraffeTypeError("Cannot determine data type")
        self.length = column.get("length")
        self.precision = column.get("precision")
        self.scale = column.get("scale")
        self._nullable = column.get("nullable", None)
        self._default = column.get("default", None)
        self._format = column.get("format", None)
        self.filler = filler
        self.order = False

    @property
    def name(self):
        if self.title:
            return self.title
        return self.original_name

    @property
    def as_tuple(self):
        return (self.safe_name, self.tpt_type, self.length,
                self.precision, self.scale)

    @property
    def default(self):
        if self._default is None:
            raise GiraffeError("Default information not available")
        return self._default

    @property
    def format(self):
        if self._format is None:
            raise GiraffeError("Default information not available")
        return self._format

    @property
    def nullable(self):
        if self._nullable is None:
            raise GiraffeError("Nullable information not available")
        return self._nullable

    @property
    def safe_name(self):
        return safe_name(self.name)

    def __eq__(self, other):
        return str(self) == str(other)

    def __repr__(self):
        type_name = COL_TYPE_NAMES.get(self.type, "unknown")
        type_name = type_name.replace('_NN', '').replace('_N' , '').lower()
        if self.type in DECIMAL_TYPES :
            type_name = "{}[{}]({},{})".format(type_name, self.length, self.precision, self.scale)
        else:
            type_name = "{}({})".format(type_name, self.length)
        options = []
        if self._default:
            options.append(self._default)
        if self._nullable:
            options.append(self._nullable)
        if self._format:
            options.append(self._format)
        if options:
            type_name = "{} {}".format(type_name, " ".join(options))
        if self.original_name:
            name = self.original_name
        elif self.alias:
            name = self.alias
        elif self.title:
            name = self.title
        else:
            name = "n/a"
        return "Column({} {})".format(safe_name(name), type_name)

    def __str__(self):
        return self.__repr__()


class Columns(object):
    """
    A set of :class:`~giraffez.types.Column` objects, used to convey
    original table schema without losing information.
    """

    __slots__ = ('columns', '_column_map', '_filtered_columns')

    def __init__(self, items=[]):
        self.columns = []
        for item in items:
            self.columns.append(Column(item))
        self._filtered_columns = []
        self._column_map = {}
        for index, column in enumerate(self.columns):
            self._column_map[column.safe_name] = index
            if column.alias:
                self._column_map[safe_name(column.alias)] = index
            if column.title:
                self._column_map[safe_name(column.title)] = index

    def append(self, item):
        if isinstance(item, Column):
            self.columns.append(item)
        else:
            self.columns.append(Column(item))

    def get(self, column_name):
        """
        Retrieve a column from the list with name value :code:`column_name`

        :param str column_name: The name of the column to get
        :return: :class:`~giraffez.types.Column` with the specified name, or :code:`None` if it does not exist.
        """
        column_name = column_name.lower()
        for c in self.columns:
            if c.name == column_name:
                return c
        return None

    @property
    def header_length(self):
        return int(math.ceil(len(self)/8.0))

    @property
    def names(self):
        """
        :return: The names of the contained :class:`~giraffez.types.Column` objects
        :rtype: list
        """
        return [c.name for c in self]

    @property
    def safe_names(self):
        return [c.safe_name for c in self]

    def set_filter(self, names=None):
        """
        Set the names of columns to be used when iterating through the list,
        retrieving names, etc.

        :param list names: A list of names to be used, or :code:`None` for all
        """
        _names = []
        if names:
            for name in names:
                _safe_name = safe_name(name)
                if _safe_name not in self._column_map:
                    raise GiraffeTypeError("Column '{}' does not exist".format(name))
                if _safe_name in _names:
                    continue
                _names.append(_safe_name)
        self._filtered_columns = _names

    def tuples(self):
        return [c.as_tuple for c in self]

    def __eq__(self, other):
        return all([a == b for a, b in zip(self, other)])

    def __getitem__(self, index):
        if self._filtered_columns:
            return self.get(self._filtered_columns[index])
        return self.columns[index]

    def __iter__(self):
        if self._filtered_columns:
            for c in self._filtered_columns:
                column = self.get(c)
                if column is None:
                    raise GiraffeTypeError("Column '{}' does not exist".format(c))
                yield column
        else:
            for c in self.columns:
                yield c

    def __len__(self):
        return len(self.names)

    def __str__(self):
        return ",".join([c.name for c in self.columns])

    def __repr__(self):
        return "\n".join([repr(c) for c in self.columns])

    def serialize(self):
        """
        Serializes the columns into the giraffez archive header
        binary format::

            0      1      2
            +------+------+------+------+------+------+------+------+
            | Header      | Header Data                             |
            | Length      |                                         |
            +------+------+------+------+------+------+------+------+

                       giraffez Archive Header Format

                                   Fig. 1

            Header Length: 2 bytes
                Full length of archive header 

            Header Data: variable
                Binary data representing N column(s) using the format
                specified in Fig. 2


            0      1      2      3      4      5      6      7      8
            +------+------+------+------+------+------+------+------+
            | Type        | Length      | Precision   | Scale       | 
            +------+------+------+------+------+------+------+------+
            | Name Length | Name                                    |
            +------+------+------+------+------+------+------+------+

                           Binary Column Format

                                   Fig. 2

            Type: 2 bytes
                Numerical representation of column type defined by
                Teradata's CLIv2

            Length: 2 bytes
                Column length

            Scale: 2 bytes
                Column scale

            Precision: 2 bytes
                Column precision

            Name Length: 2 bytes
                Length used for reading variable column name

            Name: variable
                Name of column

        :return: Packed binary data, representing the serialized :class:`~giraffez.types.Columns`
        :rtype: str
        """
        data = b""
        for column in self:
            row = struct.pack("5H", column.type, column.length, column.precision, column.scale,
                len(column.name))
            row += ensure_bytes(column.name)
            data += row
        return struct.pack("H", len(data)) + data

    @classmethod
    def deserialize(cls, data):
        """
        Deserializes giraffez Archive header. See
        :meth:`~giraffez.types.Columns.serialize` for more information.

        :param str data: data in giraffez Archive format, to be deserialized
        :return: :class:`~giraffez.types.Columns` object decoded from data
        """
        column_list = cls()
        while data:
            tup, data = data[:10], data[10:]
            column_type, length, prec, scale, title_len = struct.unpack("5H", tup)
            title, data = data[:title_len], data[title_len:]
            try:
                column_list.append((title, column_type, length, prec, scale))
            except GiraffeTypeError as error:
                raise GiraffeEncodeError(error)
        return column_list


class Date(datetime.datetime):
    """
    Ensures that datetime objects can be proper coerced into a
    string value of a given format.

    There is a Python bug (yes, it is a bug, not a feature) where
    datetime objects representing any dates before 1900 raises an
    error when calling the strftime method. This will ensure that the
    date is coerced safely with str() into something Teradata can use.
    """

    def __init__(self, *args, **kwargs):
        super(Date, self).__init__()

    @classmethod
    def from_datetime(cls, dt):
        if isinstance(dt, datetime.datetime):
            return cls(dt.year, dt.month, dt.day, dt.hour, dt.minute, dt.second)
        else:
            return cls(dt.year, dt.month, dt.day, 0, 0, 0)

    @classmethod
    def from_integer(cls, d):
        try:
            if not d:
                return None
            d += 19000000
            year = d // 10000;
            month = (d % 10000) // 100
            day = d % 100
            return cls(year=year, month=month, day=day)
        except ValueError as error:
            log.debug(error)
            return None

    @classmethod
    def from_string(cls, s):
        for fmt in DATE_FORMATS:
            try:
                return cls.strptime(str(s), fmt)
            except ValueError as error:
                pass
        return None

    def to_integer(self):
        try:
            date_value = "{0:04}{1:02}{2:02}".format(self.year, self.month, self.day)
            return int(date_value) - 19000000
        except ValueError as error:
            return None

    def __json__(self):
        return unicode(self.to_string())

    def to_string(self):
        return "{0:04}-{1:02}-{2:02}".format(self.year, self.month, self.day)

    def __bytes__(self):
        return bytes(self.to_string(), "UTF-8")

    def __int__(self):
        return self.to_integer()

    def __str__(self):
        return self.to_string()

    def __unicode__(self):
        return unicode(self.to_string())


class Decimal(decimal.Decimal):
    def __json__(self):
        return self.__str__()


class Row(object):
    """
    A wrapper for a single row object.

    Defines the :code:`__iter__` magic method for convenience:

    .. code-block:: python

       for item in row:
           print(item)

    Defines the :code:`__getattr__` magic method to return a particular field:

    .. code-block:: python

       print(row.first_name) # 'alice'

    Defines the :code:`__getitem__` magic method to return a particular field by name or numeric index:

    .. code-block:: python

       print(row['first_name']) # 'alice'
       print(row['id']) # 'abc123'
       print(row[2]) # 'abc123'
    """

    __slots__ = ('columns', 'row')

    def __init__(self, columns, row):
        self.columns = columns
        self.row = row

    def items(self):
        """
        Represents the contents of the row as a :code:`dict` with the column
        names as keys, and the row's fields as values.

        :rtype: dict
        """
        return {k.name: v for k, v in zip(self.columns, self)}

    def __getattr__(self, name):
        try:
            index = self.columns._column_map[name]
            return self.__getitem__(index)
        except KeyError:
            raise AttributeError("'{}' object has no attribute '{}'".format(
                self.__class__.__name__, name))

    def __getitem__(self, key):
        if isinstance(key, basestring):
            try:
                return self.row[self.columns._column_map[key]]
            except KeyError:
                raise AttributeError("Row has no column '{}'".format(key))
        else:
            return self.row.__getitem__(key)

    def __iter__(self):
        for item in self.row:
            yield item

    def __len__(self):
        return len(self.row)

    def __repr__(self):
        return "Row({})".format(self.items())

    def __json__(self):
        return self.items()

    def __str__(self):
        return str(self.items())


class Time(datetime.time):
    """
    Represents Teradata time data types such as TIME(n).
    """

    @classmethod
    def from_string(cls, s):
        fmt = TIME_FORMATS.get(len(s.strip()))
        if fmt is not None:
            try:
                ts = datetime.datetime.strptime(str(s), fmt)
                return cls(ts.hour, ts.minute, ts.second, ts.microsecond)
            except ValueError as error:
                log.debug("GiraffeTime: ", error)
        return None

    def to_string(self, length=None):
        hms = "{0:02}:{1:02}:{2:02}".format(self.hour, self.minute, self.second)
        if length is not None and length > 8:
            ms = "{:06}".format(self.microsecond)[:length - 9]
            return "{0}.{1}".format(hms, ms)
        return hms


class Timestamp(Date):
    """
    Represents Teradata date/time data types such as TIMESTAMP(n).
    """

    def __init__(self, *args, **kwargs):
        super(Timestamp, self).__init__(*args, **kwargs)
        self._original_length = None

    @classmethod
    def from_string(cls, s):
        s = s.strip()
        format = TIME_FORMATS.get(len(s))
        if format is not None:
            try:
                if len(s) > 26:
                    s = s[:26]
                ts = cls.strptime(str(s), format)
                ts._original_length = len(s)
                return ts
            except ValueError as error:
                log.debug("GiraffeTimestamp: ", error)
        return None

    def to_string(self, length=None):
        if length is None:
            length = self._original_length
        if length is not None and length > 20:
            value = "{0:04}-{1:02}-{2:02} {3:02}:{4:02}:{5:02}.{6:06}".format(self.year, self.month, self.day,
                    self.hour, self.minute, self.second, self.microsecond)
            value = value[:length]
            return value
        return "{0:04}-{1:02}-{2:02} {3:02}:{4:02}:{5:02}".format(self.year, self.month, self.day,
                    self.hour, self.minute, self.second)
